File: DevExpress VCL/ExpressPivotGrid 2/Demos/Delphi/OLAPBrowser/OLAPBrowserMain.pas
unit OLAPBrowserMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, cxLookAndFeels, Menus, StdCtrls, cxControls,
  cxCustomPivotGrid, cxPivotGrid, cxCustomData, cxClasses, cxGraphics,
  cxStyles, cxPivotGridOLAPConnectionDesigner, cxPivotGridOLAPDataSource,
  cxEdit;

type
  TfrmOlapBrowser = class(TfrmDemoBaisicMain)
    UnboundPivotGrid: TcxPivotGrid;
    NewConnection1: TMenuItem;
    N3: TMenuItem;
    OLAPDataSource: TcxPivotGridOLAPDataSource;
    procedure FormCreate(Sender: TObject);
    procedure NewConnection1Click(Sender: TObject);
  protected
    function GetPivotGrid: TcxCustomPivotGrid; override;
    procedure SetFieldPos(const AFieldName: string; AArea: TcxPivotGridFieldArea);
  public
    { Public declarations }
  end;

var
  frmOlapBrowser: TfrmOlapBrowser;

implementation

{$R *.dfm}


function TfrmOlapBrowser.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := UnboundPivotGrid;
end;

procedure TfrmOlapBrowser.SetFieldPos(
  const AFieldName: string; AArea: TcxPivotGridFieldArea);
var
  AField: TcxPivotGridField;
begin
  AField := PivotGrid.GetFieldByName(AFieldName);
  if AField = nil then Exit;
  AField.Area := AArea;
  AField.Visible := True;
end;

procedure TfrmOlapBrowser.FormCreate(Sender: TObject);
var
  APath, ConnectionString: string;
begin
  inherited;
  APath := ExpandFileName('..\..\Data\Northwind.cub');
  if not FileExists(APath) then Exit;
  ConnectionString := 'Provider=MSOLAP.3;Integrated Security=SSPI;Persist Security Info=False;Data Source=';
  PivotGrid.BeginUpdate;
  try
    OLAPDataSource.ConnectionString := ConnectionString + APath;
    OLAPDataSource.Active := True;
    OLAPDataSource.RetrieveFields(PivotGrid);
    SetFieldPos('Country', faColumn);
    SetFieldPos('City', faColumn);
    SetFieldPos('Category Name', faRow);
    SetFieldPos('Products', faRow);
    SetFieldPos('Quantity', faData);
    SetFieldPos('Discount', faData);
  finally
    PivotGrid.EndUpdate;
    PivotGrid.ApplyBestFit;
  end;
end;

procedure TfrmOlapBrowser.NewConnection1Click(Sender: TObject);
var
  I, J: Integer;
  ACube, ANewConnectionString: WideString;
begin
  ANewConnectionString := cxPivotGridOLAPCreateConnectionString(ACube, PivotGrid.LookAndFeel);
  if ANewConnectionString <> '' then
  begin
    PivotGrid.BeginUpdate;
    try
      OLAPDataSource.ConnectionString := ANewConnectionString;
      OLAPDataSource.Cube := ACube;
      OLAPDataSource.Active := True;
      OLAPDataSource.RetrieveFields(PivotGrid);
      for I := 0 to PivotGrid.Groups.Count - 1 do
        for J := 1 to PivotGrid.Groups[I].FieldCount - 1 do
          PivotGrid.Groups[I].Fields[J].Visible := True;
    finally
      PivotGrid.EndUpdate;
      PivotGrid.Customization.Visible := True;
    end;
  end;
end;

end.

File: DevExpress VCL/ExpressPivotGrid 2/Demos/CBuilder/OLAPBrowser/OLAPBrowserMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "OLAPBrowserMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxCustomPivotGrid"
#pragma link "cxGraphics"
#pragma link "cxPivotGrid"
#pragma link "cxPivotGridOLAPDataSource"
#pragma link "cxPivotGridOLAPConnectionDesigner"
#pragma link "cxStyles"
#pragma resource "*.dfm"
TfrmOlapBrowser *frmOlapBrowser;
//---------------------------------------------------------------------------
__fastcall TfrmOlapBrowser::TfrmOlapBrowser(TComponent* Owner)
	: TfrmDemoBaisicMain(Owner)
{
}
//---------------------------------------------------------------------------

TcxCustomPivotGrid* __fastcall TfrmOlapBrowser::PivotGrid()
{
  return UnboundPivotGrid;
}
//---------------------------------------------------------------------------

void __fastcall TfrmOlapBrowser::SetFieldPos(
  const String AFieldName, TcxPivotGridFieldArea AArea)
{
  TcxPivotGridField* AField = PivotGrid()->GetFieldByName(AFieldName);
  AField->Area = AArea;
  AField->Visible = true;
}

void __fastcall TfrmOlapBrowser::FormCreate(TObject *Sender)
{
  TfrmDemoBaisicMain::FormCreate(Sender); 
  WideString APath = ExpandFileName("..\\..\\Data\\Northwind.cub");
  if (!FileExists(APath)) return;
  WideString ConnectionString = "Provider=MSOLAP.3;Integrated Security=SSPI;Persist Security Info=False;Data Source=";
  PivotGrid()->BeginUpdate();
  try {
	OLAPDataSource->ConnectionString = ConnectionString + APath;
	OLAPDataSource->Active = true;
	OLAPDataSource->RetrieveFields(PivotGrid());
	SetFieldPos("Country", faColumn);
	SetFieldPos("City", faColumn);
	SetFieldPos("Category Name", faRow);
	SetFieldPos("Products", faRow);
	SetFieldPos("Quantity", faData);
	SetFieldPos("Discount", faData);
  }
  __finally {
	PivotGrid()->EndUpdate();
	PivotGrid()->ApplyBestFit();
  }
}

void __fastcall TfrmOlapBrowser::NewConnection1Click(TObject *Sender)
{
  int I, J;
  WideString ACube;

  PivotGrid()->BeginUpdate();
  try {
	OLAPDataSource->ConnectionString =
	  cxPivotGridOLAPCreateConnectionString(ACube, PivotGrid()->LookAndFeel);
	OLAPDataSource->Cube = ACube;
	OLAPDataSource->Active = True;
	OLAPDataSource->RetrieveFields(PivotGrid());
	for (I = 0; I < PivotGrid()->Groups->Count - 1; I++) {
	  for (J = 1; J < PivotGrid()->Groups->Items[I]->FieldCount; J++) {
		PivotGrid()->Groups->Items[I]->Fields[J]->Visible = true;
	  }
	}
  }
  __finally {
	PivotGrid()->EndUpdate();
	PivotGrid()->Customization->Visible = true;
  };
}
